File: win-prog/ProgWinCS/Font Fun/HollowFontWidePen/HollowFontWidePen.cs
//------------------------------------------------
// HollowFontWidePen.cs © 2001 by Charles Petzold
//------------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

class HollowFontWidePen: FontMenuForm
{
     public new static void Main()
     {
          Application.Run(new HollowFontWidePen());
     }
     public HollowFontWidePen()
     {
          Text = "Hollow Font (Wide Pen)";
          Width *= 2; 
          strText = "Wide Pen";
          font = new Font("Times New Roman", 108, 
                          FontStyle.Bold | FontStyle.Italic);
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          GraphicsPath path = new GraphicsPath();
          float fFontSize = PointsToPageUnits(grfx, font);

               // Add text to the path.

          path.AddString(strText, font.FontFamily, (int) font.Style,
                         fFontSize, new PointF(0, 0), new StringFormat());

               // Get the path bounds for centering.

          RectangleF rectfBounds = path.GetBounds();
          
          grfx.TranslateTransform(
                         (cx - rectfBounds.Width) / 2 - rectfBounds.Left,
                         (cy - rectfBounds.Height) / 2 - rectfBounds.Top);

               // Draw the path.

          Brush brush = new HatchBrush(HatchStyle.Trellis, 
                                       Color.White, Color.Black);
          Pen pen = new Pen(brush, fFontSize / 20);
          grfx.DrawPath(pen, path);
     }
}

File: win-prog/ProgWinVB/Font Fun/HollowFontWidePen/HollowFontWidePen.vb
'--------------------------------------------------
' HollowFontWidePen.vb (c) 2002 by Charles Petzold
'--------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Class HollowFontWidePen
    Inherits FontMenuForm

    Shared Shadows Sub Main()
        Application.Run(New HollowFontWidePen())
    End Sub

    Sub New()
        Text = "Hollow Font (Wide Pen)"
        Width *= 2
        strText = "Wide Pen"
        fnt = New Font("Times New Roman", 108, _
                       FontStyle.Bold Or FontStyle.Italic)
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Dim path As New GraphicsPath()
        Dim fFontSize As Single = PointsToPageUnits(grfx, fnt)

        ' Add text to the path.
        path.AddString(strText, fnt.FontFamily, fnt.Style, _
                       fFontSize, New PointF(0, 0), New StringFormat())

        ' Get the path bounds for centering.
        Dim rectfBounds As RectangleF = path.GetBounds()
        grfx.TranslateTransform( _
                    (cx - rectfBounds.Width) / 2 - rectfBounds.Left, _
                    (cy - rectfBounds.Height) / 2 - rectfBounds.Top)

        ' Draw the path.
        Dim br As New HatchBrush(HatchStyle.Trellis, _
                                 Color.White, Color.Black)
        Dim pn As New Pen(br, fFontSize / 20)
        grfx.DrawPath(pn, path)
    End Sub
End Class
